src/protocol/sort/shuffle.rs

```diff
@@ -74,127 +67,147 @@ pub fn get_two_of_three_random_permutations(
+            ShuffleOrUnshuffle::Shuffle => apply_inv(permutation_to_apply, input),
+            ShuffleOrUnshuffle::Unshuffle => apply(permutation_to_apply, input),
         }
-        self.reshare_all_shares(&ctx, to_helper).await
     }
+    reshare_all_shares(input, &ctx, to_helper).await
+}
 
-    #[embed_doc_image("shuffle", "images/sort/shuffle.png")]
-    /// Shuffle calls `shuffle_or_unshuffle_once` three times with 2 helpers shuffling the shares each time.
-    /// Order of calling `shuffle_or_unshuffle_once` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
-    /// Each shuffle requires communication between helpers to perform reshare.
-    /// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
-    /// For this, we have three shuffle steps one per `shuffle_or_unshuffle_once` i.e. Step1, Step2 and Step3.
-    /// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
-    /// ![Shuffle steps][shuffle]
-    pub async fn execute(
-        &mut self,
-        ctx: ProtocolContext<'_, Replicated<F>, F>,
-    ) -> Result<Vec<Replicated<F>>, BoxError>
-    where
-        F: Field,
-    {
-        self.input = self
-            .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step1)
-            .await?;
-        self.input = self
-            .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step2)
-            .await?;
-        self.shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step3)
-            .await
-    }
+#[embed_doc_image("shuffle", "images/sort/shuffle.png")]
+/// Shuffle calls `shuffle_or_unshuffle_once` three times with 2 helpers shuffling the shares each time.
+/// Order of calling `shuffle_or_unshuffle_once` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
+/// Each shuffle requires communication between helpers to perform reshare.
+/// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
+/// For this, we have three shuffle steps one per `shuffle_or_unshuffle_once` i.e. Step1, Step2 and Step3.
+/// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
+/// ![Shuffle steps][shuffle]
+pub async fn shuffle_shares<F: Field>(
+    input: &mut [Replicated<F>],
+    permutation_left: &[usize],
+    permutation_right: &[usize],
+    ctx: ProtocolContext<'_, Replicated<F>, F>,
+) -> Result<Vec<Replicated<F>>, BoxError> {
+    let mut once_shuffled = shuffle_or_unshuffle_once(
+        input,
+        permutation_left,
+        permutation_right,
+        ShuffleOrUnshuffle::Shuffle,
+        &ctx,
+        Step1,
+    )
+    .await?;
+    let mut twice_shuffled = shuffle_or_unshuffle_once(
+        &mut once_shuffled,
+        permutation_left,
+        permutation_right,
+        ShuffleOrUnshuffle::Shuffle,
+        &ctx,
+        Step2,
+    )
+    .await?;
+    shuffle_or_unshuffle_once(
+        &mut twice_shuffled,
```

Comment:
At this point, we will have input, once_shuffled and twice_shuffled in the memory since none of them is dropped.

Reply:
That's true - and it's not ideal. Let me see if I can improve it.